zellij-server/src/os_input_output.rs

```diff
@@ -97,96 +100,63 @@ fn handle_command_exit(mut child: Child) {
     }
 }
 
-fn handle_fork_pty(
-    fork_pty_res: ForkptyResult,
+fn handle_openpty(
+    open_pty_res: OpenptyResult,
     cmd: RunCommand,
-    parent_fd: RawFd,
-    child_fd: RawFd,
-) -> (RawFd, ChildId) {
-    let pid_primary = fork_pty_res.master;
-    let (pid_secondary, pid_shell) = match fork_pty_res.fork_result {
-        ForkResult::Parent { child } => {
-            let pid_shell = read_from_pipe(parent_fd, child_fd);
-            (child, pid_shell)
+    quit_cb: Box<dyn Fn(PaneId) + Send>,
+) -> (RawFd, RawFd) {
+    // primary side of pty and child fd
+    let pid_primary = open_pty_res.master;
+    let pid_secondary = open_pty_res.slave;
+
+    let mut child = unsafe {
+        let command = &mut Command::new(cmd.command);
+        if let Some(current_dir) = cmd.cwd {
+            command.current_dir(current_dir);
         }
-        ForkResult::Child => {
-            let child = unsafe {
-                let command = &mut Command::new(cmd.command);
-                if let Some(current_dir) = cmd.cwd {
-                    command.current_dir(current_dir);
+        command
+            .args(&cmd.args)
+            .pre_exec(move || -> std::io::Result<()> {
+                if libc::login_tty(pid_secondary) != 0 {
+                    panic!("failed to set controlling terminal");
```

Comment:
I'm not sure if panics are safe in pre_exec. You can return some io error, and catch the error code (but not the text) in parent process as far as I remember.